src/macros.rs

```diff
@@ -65,3 +65,25 @@ macro_rules! to_hex_string {
         format!("0x{}", output)
     }};
 }
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! core_uri {
+    ($info_provider:tt, $chain_id:expr, $path:literal) => {{
+        let result: ApiResult<String> =
+        match $info_provider.chain_info($chain_id).await {
+            Ok(chain_info) => Ok(format!("{}{}",chain_info.tx_service_url, $path)),
+            Err(error) => Err(error,)
+        };
+        result
+    }};
+    ($info_provider:tt, $chain_id:expr, $path:literal, $($arg:tt)*) => {{
+        let full_path = format!($path, $($arg)*);
+        let result: ApiResult<String> =
+        match $info_provider.chain_info($chain_id).await {
```

Comment:
Why can we not use `?`?

Reply:
Becauses `?` can only be used in methods that return `Result<V, E>` , which in our case it is most of the time, except for unit tests. By `match`ing the `result` we are writing the macro in a way that makes no assumption on the calling context.